File: python/102/patterns/pipelinepattern.py
from typing import Callable, List

# Pipeline Step Type
PipelineStep = Callable[[str], str]

class Pipeline:
    def __init__(self):
        self.steps: List[PipelineStep] = []

    def add_step(self, step: PipelineStep) -> "Pipeline":
        self.steps.append(step)
        return self  # umożliwia łańcuchowe dodawanie

    def execute(self, input: str) -> str:
        result = input
        for step in self.steps:
            result = step(result)
        return result

# Funkcje reprezentujące kroki pipeline
def trim(input: str) -> str:
    return input.strip()

def to_uppercase(input: str) -> str:
    return input.upper()

def add_prefix(prefix: str) -> Callable[[str], str]:
    def step(input: str) -> str:
        return f"{prefix}{input}"
    return step

# Tworzymy pipeline
text_pipeline = Pipeline() \
    .add_step(trim) \
    .add_step(to_uppercase) \
    .add_step(add_prefix("Processed: "))

# Przykład użycia
result = text_pipeline.execute("   hello world   ")
print(result)  # "Processed: HELLO WORLD"

File: typescript-basic/classes/pipelinepattern.ts
type PipelineStep = (input: string) => string;

// Klasa Pipeline
class Pipeline {
  private steps: PipelineStep[] = [];

  // Dodajemy krok do pipeline
  addStep(step: PipelineStep): this {
    this.steps.push(step);
    return this; // umożliwia łańcuchowe dodawanie
  }

  // Uruchamiamy pipeline na danych wejściowych
  execute(input: string): string {
    return this.steps.reduce((data, step) => step(data), input);
  }
}

// Funkcje reprezentujące kroki pipeline
const trim = (input: string): string => input.trim();
const toUpperCase = (input: string): string => input.toUpperCase();
const addPrefix = (prefix: string) => (input: string): string => `${prefix}${input}`;

// Tworzymy pipeline
const textPipeline = new Pipeline()
  .addStep(trim)
  .addStep(toUpperCase)
  .addStep(addPrefix("Processed: "));

// Przykład użycia
const result = textPipeline.execute("   hello world   ");
console.log(result); // "Processed: HELLO WORLD"